### ISSUE TITLE
Changing base architecture for SSD300

### ISSUE THREAD

**Issue by f4br1** [pierluigiferrari/ssd_keras#88]
Hi,

I am trying to use a different architecture than VGG16. I am planning to use Mobilenets as base architecture. Do I have to add the "Extra Feature Layer" on top of Mobilenets as the SSD paper does for VGG16 or how does one proceed by changing the base architecture?

Any suggestions will be greatly appreciated.

Thanks!

**Comment by pierluigiferrari** [comment]
If by "extra feature layers" you mean what is `conv6` to `conv9` in SSD300, then no, you don't have to add those to MobileNet. You can add extra convolutional layers though. It's your design choice, really. The point is that you choose your base network, then you decide from which layers of the base network you want to make predictions, and then you add predictor layers as described in `keras_ssd7.py` on top of those respective layers in the base network.

Whether or not you want to add additional layers on top of MobileNet also depends on what spatial resolution you want for your last predictor layer(s). For example, the extra feature layers on top of the VGG16 in SSD300 result in the output feature map of `conv9_2` having spatial dimensions exactly 1x1, so the last predictor layer predicts only one large box spatially. You may or may not want to create the same design for your MobileNet SSD. As I said, that's your design choice.

**Comment by f4br1** [comment]
Thank you!. That clarified my question. 

About the "extra feature layers" , yes, I meant the convolutional layers conv6 to conv11. I will add the predictor layers for some of the base network layers.

I have two more questions though. 

1.) Do I also have to use pre-trained weights for Mobilenets too as you do for VGG16? 

2.) How can I compute the mean average precision for the detection problem?  Is there any keras function or similar? Say that I want to compute mAP for the examples that you have on your tutorial.

**Comment by pierluigiferrari** [comment]
1. I mean, is there any reason not to use pre-trained MobileNet weights? They are right there in `keras.Applications`, you only need to download them. If MobileNet can be trained end to end from scratch (i.e without any pre-trained weights), then a MobileNet SSD can probably be trained end to end from scratch, too. If not, then probably not, although it cannot be said with certainty, you would just have to try.
2. There is no universal definition of the mAP. It's a much more complex metric than something like classification accuracy and there are several variations of it. The authors of the Pascal VOC competitions used one formula for their 2007 competition, and then a slightly different formula for their 2012 competition. In the MS COCO competitions, yet another formula is used. They are all similar, but not identical. So if you want to compute the mAP, you have to decide what version of it you want to compute first. If you want to compute the MS COCO-style mAP, then you can use the `pycoco` tools that come with COCO. If you want to compute the Pascal VOC-style mAP, then you can either use their official Matlab scripts or find some Python implementation somewhere if you can. Or write your own, of course. In any case you can always use the evaluation notebooks in this repo to produce the prediction output in the format that is required by the aforementioned mAP calculators.
